packages/protobuf/src/binary-encoding.ts

```diff
@@ -329,15 +331,28 @@ export class BinaryWriter implements IBinaryWriter {
       bytes.set(this.chunks[i], offset);
       offset += this.chunks[i].length;
     }
-    this.chunks = [];
+
+    this.restore();
```

Comment:
The documentation states that the method "resets this writer". It looks like the function does not reset the property `buf` . It makes sense to me to fix this up, and add `this.buf = []` here. 

From what I see, this change would be sufficient for your use case, and it would improve the behavior of the writer. I do not think that any other changes should be made for this fix, at least not without adding significant test coverage first and taking a close look at the impact on performance.